internal/util/importutilv2/numpy/column_reader.go

```diff
@@ -0,0 +1,296 @@
+		data, err := ReadN[uint8](c.reader, c.order, count)
+		if err != nil {
+			return nil, err
+		}
+		c.readPosition += int(count)
+		return &storage.BinaryVectorFieldData{
+			Data: data,
+			Dim:  int(c.dim),
+		}, nil
+	case schemapb.DataType_FloatVector:
+		elementType, err := convertNumpyType(c.npyReader.Header.Descr.Type)
+		if err != nil {
+			return nil, err
+		}
+		var data []float32
+		switch elementType {
+		case schemapb.DataType_Float:
+			data, err = ReadN[float32](c.reader, c.order, count)
+			if err != nil {
+				return nil, err
+			}
+			err = typeutil.VerifyFloats32(data)
+			if err != nil {
+				return nil, nil
+			}
+		case schemapb.DataType_Double:
+			data64, err := ReadN[float64](c.reader, c.order, count)
+			if err != nil {
+				return nil, err
+			}
+			err = typeutil.VerifyFloats64(data64)
+			if err != nil {
+				return nil, err
+			}
+			data = lo.Map(data64, func(f float64, _ int) float32 {
+				return float32(f)
+			})
+		}
+		c.readPosition += int(count)
+		return &storage.FloatVectorFieldData{
+			Data: data,
+			Dim:  int(c.dim),
+		}, nil
+	default:
+		return nil, merr.WrapErrImportFailed(fmt.Sprintf("unsupported data type: %s", dt.String()))
+	}
+}
+
+func (c *ColumnReader) Close() {}
+
+// setByteOrder sets BigEndian/LittleEndian, the logic of this method is copied from npyio lib
+func (c *ColumnReader) setByteOrder() {
+	var nativeEndian binary.ByteOrder
+	v := uint16(1)
+	switch byte(v >> 8) {
+	case 0:
+		nativeEndian = binary.LittleEndian
+	case 1:
+		nativeEndian = binary.BigEndian
+	}
+
+	switch c.npyReader.Header.Descr.Type[0] {
+	case '<':
+		c.order = binary.LittleEndian
+	case '>':
+		c.order = binary.BigEndian
+	default:
+		c.order = nativeEndian
+	}
+}
+
+func (c *ColumnReader) ReadString(count int64) ([]string, error) {
+	// varchar length, this is the max length, some item is shorter than this length, but they also occupy bytes of max length
+	maxLen, utf, err := stringLen(c.npyReader.Header.Descr.Type)
+	if err != nil || maxLen <= 0 {
+		return nil, merr.WrapErrImportFailed(
+			fmt.Sprintf("failed to get max length %d of varchar from numpy file header, error: %v", maxLen, err))
+	}
+
+	// read data
+	data := make([]string, 0, count)
+	for len(data) < int(count) {
+		if utf {
+			// in the numpy file with utf32 encoding, the dType could be like "<U2",
+			// "<" is byteorder(LittleEndian), "U" means it is utf32 encoding, "2" means the max length of strings is 2(characters)
+			// each character occupy 4 bytes, each string occupies 4*maxLen bytes
+			// for example, a numpy file has two strings: "a" and "bb", the maxLen is 2, byte order is LittleEndian
+			// the character "a" occupies 2*4=8 bytes(0x97,0x00,0x00,0x00,0x00,0x00,0x00,0x00),
+			// the "bb" occupies 8 bytes(0x97,0x00,0x00,0x00,0x98,0x00,0x00,0x00)
+			// for non-ascii characters, the unicode could be 1 ~ 4 bytes, each character occupies 4 bytes, too
+			raw, err := io.ReadAll(io.LimitReader(c.reader, utf8.UTFMax*int64(maxLen)))
```

Comment:
Does "ReadAll" here read the entire content of the file?

Reply:
nope